### PR TITLE
Fast5

### LINKED ISSUE #80: Fast5 upload support
## Describe your idea for a new feature
Add a parser for uploading fast5 data. Need details on what the data structure is going to look like before development.

## Additional information
tbd

### PR DESCRIPTION
**Pull request by JeffreyThiessen**
## Description of changes
* Added support for uploading `fast5` files
* Added `--upload_mode=<mode>` as an argument, replacing the `--assemblies` argument
  * Supports the following upload modes, which each upload files to different IRIDA REST endpoints
    * `default`: Sends to `sequencingRuns`, this is the default behaviour
    * `assemblies`: Sends to `assemblies`, this replaces `--assemblies`
    * `fast5`: Sends to `fast5`, this is a new feature
    * GUI now has a dropdown box to select upload mode

## Related issue
Fixes #80 

## Checklist
Things for the developer to confirm they've done before the PR should be accepted:

* [x] CHANGELOG.md updated with information for new change.
* [x] Tests added (or description of how to test) for any new features.
* [x] User documentation updated for UI or technical changes.

### CODE REVIEW

File: iridauploader/model/sequencing_run.py [hunk 1]
```diff
@@ -34,20 +34,15 @@ class SequencingRun:
                 }
             }
         },
-        '_assemblies': {
-            'type': 'boolean',
-            'required': True
-        },
```
Comment (tom114) [inline iridauploader/model/sequencing_run.py:40]: Was this just an unused property on the sequencing run?  No need for adding the upload mode here?

File: iridauploader/api/api_calls.py [hunk 2]
```diff
@@ -20,6 +20,17 @@
 from . import exceptions
 
 
+MODE_DEFAULT = "default"
+MODE_ASSEMBLIES = "assemblies"
+MODE_FAST5 = "fast5"
+
+UPLOAD_MODES = [
```
Comment (tom114) [inline iridauploader/api/api_calls.py:27]: Add a comment in here above these stating what they're for please.

File: iridauploader/api/api_calls.py [hunk 3]
```diff
@@ -500,6 +511,52 @@ def get_assemblies_files(self, project_id, sample_name):
...
+                                 })
+            response = self._session.get(url)
+
+        except StopIteration:
+            logging.error("The given sample doesn't exist: ".format(sample_name))
+            raise exceptions.IridaResourceError("The given sample ID doesn't exist", sample_name)
+
+        # todo future development if needed one day
+        # This response should be returned as some sort of file object
+        # This is related to how we return get_sequence_files too, but there is no real use for it at the moment, yagni
```
Comment (tom114) [inline iridauploader/api/api_calls.py:555]: `yagni` ah my favourite acronym, thanks for implementing the method though!

File: iridauploader/model/sequencing_run.py (hunk 1 above)
Comment (JeffreyThiessen) [reply iridauploader/model/sequencing_run.py:40]: Previously when I implemented assemblies, My thought process was that if the files on a `SequencingRun` were assemblies, we should denote that on the object, s.t. when it uploads, the `api_handler` queries the `SequencingRun` for file type, and chooses the appropriate api call to make.

Now that we have `default` (illumina), `assemblies`, and `fast5` files, I rethought how it should be handled, and decided the code flow would be cleaner/more logical to remove the additional steps of appending a file type to the `SequencingRun` based off of user input, and immediately fetching it again in the handler, and instead have the user option go directly to the handler. 

The reason this is cleaner is because it removes Model dependent logic from the `api` layer, and leaves the api to just do api things. Additionally it simplifies the handler logic, so it doesn't need to append/fetch variables to/from the model when it already has the data on the same layer.

In short: Made the simple solution with the assemblies branch because it was a single case. Now that it's more complex, refactor to make it generic.

File: iridauploader/api/api_calls.py (hunk 2 above)
Comment (JeffreyThiessen) [reply iridauploader/api/api_calls.py:29]: Added.

File: iridauploader/model/sequencing_run.py (hunk 1 above)
Comment (tom114) [reply iridauploader/model/sequencing_run.py:40]: Sounds good, thanks for the full description.

File: CHANGELOG.md [hunk 4]
```diff
@@ -5,6 +5,13 @@ Beta 0.5.0
 ----------
 Added functionality:
 * Added `nanopore_assemblies` as an alias for directory uploads
+* Added support for uploading `fast5` files
+* Added `--upload_mode=<mode>` as an argument, replacing the `--assemblies` argument
```
Comment (tom114) [inline CHANGELOG.md:9]: Still need this information added to the main docs that get published on readthedocs

Comment (JeffreyThiessen) [reply CHANGELOG.md:9]: Added.

Comment (tom114) [reply CHANGELOG.md:9]: I still see a `--assemblies` section in https://irida-uploader.readthedocs.io/en/stable/index.html#assemblies-uploading

Comment (JeffreyThiessen) [reply CHANGELOG.md:9]: Oops, committing too fast, fixed now

**Comment by tom114** [comment]
Tested and this appears to work well for `.fast5` and zipped `.fast5.tar.gz` files.  Thanks. Merging.
